src/types/input.jl

```diff
@@ -2,81 +2,151 @@
 
 abstract type InputType end
 
-struct Bosonic <:InputType
-end
-struct PartDist <:InputType
-end
-abstract type ToyModel <: InputType
-end
-abstract type RandomModel <: InputType
-end
-struct Distinguishable <:InputType
+abstract type Bosonic <: InputType end
+
+abstract type GaussianState <: Bosonic end
+
+struct ThermalState <: GaussianState
+
+    mode_number::Int
+    mean_photon_number::Float64
+    mean_vector::Vector
+    covariance_matrix::Matrix
+
+    function ThermalState(mode_number::Int, mean_photon_number::Float64)
+        return new(
+            mode_number,
+            mean_photon_number,
+            zeros(mode_number),
+            (mean_photon_number + 1 / 2) *
+            Matrix{ComplexF64}(I, 2*mode_number, 2*mode_number),
+        )
+    end
+
 end
-struct Undef <:InputType
+
+abstract type PartDist <: InputType end
+
+struct ToyModel <: PartDist
+    distinguishability::Float64
+    ToyModel(distinguishability::Float64) = new(distinguishability)
 end
 
+abstract type RandomModel <: PartDist end
+
+abstract type Undef <: InputType end
+
+abstract type Distinguishable <: InputType end
+
 mutable struct OrthonormalBasis
+
     """basis of vectors v_1,...,v_n stored as columns in a n*r matrix
     possibly empty"""
+
     vectors_matrix::Union{Matrix,Nothing}
+
     function OrthonormalBasis(vectors_matrix = nothing)
         if vectors_matrix == nothing
             new(nothing)
         else
             is_orthonormal(vectors_matrix, atol = ATOL) ? new(vectors_matrix) : error("invalid orthonormal basis")
         end
     end
+
 end
 
-struct GramMatrix{T<:InputType}
+mutable struct GramMatrix{T<:InputType}
```

Comment:
I'm not sure putting stuff mutable for this is good for performance, the gram matrix shouldn't change anyways and the previous code accounted that